### PR TITLE
Autocomplete: support asynchronous options

### LINKED ISSUE #87521: [Autocomplete] rebuild options when the widget configuration changes
## Use case

In short, make `Autocomplete` compatible with asynchronous completion option sources (network, DB, ...)

If completions are not immediately available as the user types, there is currently no way to feed in the new completions once they become available. The `Autocomplete` widget does not call `optionsBuilder` when the widget configuration changes, or is missing a mechanism to manually trigger a rebuild of the options.

Consider the following example:
```dart
import 'package:flutter/material.dart';

void main() => runApp(const MaterialApp(home: MyHomePage()));

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  var _options = <String>[];

  Future<void> _asyncFetchOptions(String input) async {
    // pretend to fetch the options from a remote server
    setState(() => _options = input.split(''));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Autocomplete')),
      body: Autocomplete<String>(
        fieldViewBuilder: (context, controller, focusNode, onSubmitted) {
          return TextFormField(
            autofocus: true,
            focusNode: focusNode,
            controller: controller,
            onFieldSubmitted: (_) => onSubmitted(),
            onChanged: _asyncFetchOptions,
          );
        },
        optionsBuilder: (value) => _options,
      ),
    );
  }
}
```

The result is that options are always one step behind:

## Proposal

`RawAutocompleteState.didUpdateWidget` could update the options if the they are visible, similarly how it already updates the overlay.

### PR DESCRIPTION
**Pull request by jpnurmi**
This change adds support for asynchronous options for use cases where new options are not immediately available as the user types.

Fixes: #87521

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by justinmc — COMMENTED** [review]
What does your asynchronous usage of Autocomplete look like after this PR?  Do you make an async call inside of optionsBuilder?

I always wanted to provide async support for Autocomplete, but I want to make sure this is the cleanest way to do it.  I was thinking something like making optionsBuilder return a FutureOr, but I'm curious if that would be easier for you or not.

File: packages/flutter/test/widgets/autocomplete_test.dart [hunk 1]
```diff
@@ -647,6 +647,52 @@ void main() {
...
+
+    // Enter text to show the options.
+    await tester.enterText(find.byKey(fieldKey), 'go');
+    await tester.pumpAndSettle();
+    expect(find.byKey(optionsKey), findsOneWidget);
+    expect(lastOptions, <String>['dingo', 'flamingo', 'goose']);
+
+    // Rebuild the options while they are visible.
+    availableOptions.add('gorilla');
+    await tester.pumpWidget(builder());
```
Comment (justinmc) [inline packages/flutter/test/widgets/autocomplete_test.dart:691]: I checked and this test does fail on master, but I'm worried that it doesn't exactly test how people will use this asynchronously.

**Comment by jpnurmi** [comment]
> What does your asynchronous usage of Autocomplete look like after this PR? Do you make an async call inside of optionsBuilder?

There's a dummy example in #87521. `TextFormField.onChanged` triggers an async operation to fetch options, and `optionsBuilder` returns the options that have been asynchronously fetched.

> I was thinking something like making optionsBuilder return a FutureOr, but I'm curious if that would be easier for you or not.

That sounds like a cleaner and easier to use solution... :+1:

File: packages/flutter/lib/src/widgets/autocomplete.dart [hunk 2]
```diff
@@ -750,8 +752,8 @@ class _RawAutocompleteState<T extends Object> extends State<RawAutocomplete<T>>
   }
 
   // Called when _textEditingController changes.
-  void _onChangedField() {
-    final Iterable<T> options = widget.optionsBuilder(
+  Future<void> _onChangedField() async {
+    final Iterable<T> options = await widget.optionsBuilder(
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/autocomplete.dart:756]: This looks great, is it really this easy?  I want to make sure these cases are easy too:

  1. What if the async call fails?  Would it be easy to show an error state, or another option, to show the previous options instead?
  2. Would it be easy to show a loading state?
  2. What if an async call finishes out of order?  Similarly, what if the developer wants to cancel/ignore a pending call when the input changes?
  3. What if you want to debounce the async calls?

Maybe these are the responsibility of the developer and not the framework...  I want to make sure they're easy to handle though.  What do you think?

File: packages/flutter/test/widgets/autocomplete_test.dart [hunk 3]
```diff
@@ -647,6 +647,52 @@ void main() {
...
+    late FocusNode focusNode;
+    late TextEditingController textEditingController;
+    late Iterable<String> lastOptions;
+
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Scaffold(
+          body: RawAutocomplete<String>(
+            optionsBuilder: (TextEditingValue textEditingValue) async {
+              return kOptions.where((String option) {
```
Comment (justinmc) [inline packages/flutter/test/widgets/autocomplete_test.dart:662]: Could you add another test that returns a Future.delayed here?  I want to test that actually getting the options after a delay will work as expected.

I think you could also test the "out of order" case that I mentioned in my other comment too, if we think that's a useful test to have.  Like type in the field twice, and the first time return a Future.delayed with a long delay and the second time return quickly/instantly.

**Review by justinmc — COMMENTED** [review]
Some more advanced cases to think about in my comments.  Are you already doing any of these things in your own app I wonder?

File: packages/flutter/lib/src/widgets/autocomplete.dart (hunk 2 above)
Comment (jpnurmi) [reply packages/flutter/lib/src/widgets/autocomplete.dart:756]: 3. The "out of order" case is interesting. What would be the expected result? I guess the developer should be in charge of canceling ongoing operations, but I'm not sure what would be a nice way to indicate it to the framework that a previous call should be ignored.

Comment (justinmc) [reply packages/flutter/lib/src/widgets/autocomplete.dart:756]: Well the expected result from the user's perspective would definitely be that the options reflect the current state of the input field.  I think the developer could either cancel old requests, or they could compare the state of the input field at the time a request was sent and when it returns.

I think I'm pretty convinced that all of my points above are the responsibility of the developer, and they're all solvable with this simple FutureOr setup.  It's going to be tempting for developers to just call their API directly in optionsBuilder, but really they should have some sort of async abstraction over it that handles all of the points above.  Maybe if I get some time I can write some examples or something.

File: packages/flutter/test/widgets/autocomplete_test.dart [hunk 4]
```diff
@@ -647,6 +647,56 @@ void main() {
...
+    late TextEditingController textEditingController;
+    Iterable<String>? lastOptions;
+    const delay = Duration(milliseconds: 500);
+
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Scaffold(
+          body: RawAutocomplete<String>(
+            optionsBuilder: (TextEditingValue textEditingValue) async {
+              return Future.delayed(delay, () => kOptions.where((String option) {
```
Comment (justinmc) [inline packages/flutter/test/widgets/autocomplete_test.dart:663]: This looks good.  I was thinking we could have both tests, but I think it's fine like this, this is the most important thing to test.

File: packages/flutter/test/widgets/autocomplete_test.dart [hunk 5]
```diff
@@ -647,6 +647,56 @@ void main() {
     );
   });
 
+  testWidgets('support asynchronous options builder', (WidgetTester tester) async {
+    final GlobalKey fieldKey = GlobalKey();
+    final GlobalKey optionsKey = GlobalKey();
+    late FocusNode focusNode;
+    late TextEditingController textEditingController;
+    Iterable<String>? lastOptions;
+    const delay = Duration(milliseconds: 500);
```
Comment (justinmc) [inline packages/flutter/test/widgets/autocomplete_test.dart:656]: I think you need a type here for `delay`.

**Review by justinmc — COMMENTED** [review]
Looks like you have an analyzer failure.  I caught one analyzer problem in my comment but there might be more.

File: packages/flutter/test/widgets/autocomplete_test.dart (hunk 3 above)
Comment (jpnurmi) [reply packages/flutter/test/widgets/autocomplete_test.dart]: Should I add the "out of order" case even if it wouldn't work as expected? As in, when the delayed request returns, it overrides the second quick request.

**Review by justinmc — APPROVED** [review]
LGTM, thanks for the changes.  I'm going to try to run the Google tests on this before merge in case the change to FutureOr type could break anything.

**Comment by justinmc** [comment]
Google tests look good, this will be merged when the tree is green.

**Pull request merged**